noflushd-2.8/tmp/usr/share/man/man8/noflushd.8
.\" -*- nroff -*-
.TH noflushd 8 "Version 2.7.4 - Aug 2004"
.SH NAME
noflushd \- userland daemon to spin down idle disks.
.SH SYNOPSIS
.B noflushd -[dhirtvV] [<optargs>] [<disk>] [ -t <optargs> ] [ <disks> ... ]
.SH DESCRIPTION
.B noflushd
monitors disk activity and spins down <disk>s that have been idle for a certain
amount of time.  The <disk> argument has to be a device node to a
raw disk like, e. g. /dev/hda.  It must not be a device node to a partition
inside the disk, e. g. /dev/hda1.  Unless at least one <disk> argument is
given, noflushd tries to monitor all disks on the system as listed in
/proc/partitions using the default timeout list.  Some of the options
take additional arguments <optargs>.  See below for a complete list.
.PP
.B noflushd
actually used to live in some versions of bdflush, but since kernel version
2.2.11, bdflush had been completely replaced by the kernel-level kupdate
daemon, which was renamed into kupdated, and finally replaced again by the
pdflush daemon in the Linux 2.5 kernel series.
.PP
.SH "COMMAND\-LINE OPTIONS"
.TP 0.5i
.B "-h "
Print the usage message (help).
.TP 0.5i
.B "-v "
Print information on monitored disks upon startup.
Log information about events such as disk spinup to syslog.
Combined with -d option, produce extra verbose debugging messages.
.TP 0.5i
.B "-d "
Don't detach daemon.  Print additional debugging information.
.TP 0.5i
.B "-V "
Display version information.
.TP 0.5i
.B "-r <disk>"
Mark <disk> as read-only, meaning that noflushd will make no attempt to
sync <disk> even if it's still spinning.  Usually you'll need this option for
CD-ROMs or other read-only media that are reported in /proc/partitions.
(noflushd detects IDE CD-ROMs itself.  SCSI CD-ROMs aren't listed in
/proc/partitions so just assume you don't need this option for a start.)
.TP 0.5i
.B "-n <timeout-list>"
Set default timeouts.  Can be overridden for a single disk using the
.B "-t "
option.  See below for how to specify the timeout values.
.TP 0.5i
.B "-t <timeout-list>"
Send disk to sleep after timeout minutes have passed
without read activity.  Unlike
.BR "-n" ,
this option only applies to the next <disk> argument following on the
command-line.  See below for how to specify the timeout values.
.TP 0.5i
.B "<timeout-list>"
A <timeout-list> comprises of one or more comma-separated integer values
indicating a timeout value in minutes.  Usually you'll only supply a single
value and ignore the rest of this section.  Read on however if you want to 
pass several configurations to noflushd.  Upon startup, noflushd uses the
first timeout value of each list.  When a
.I SIGHUP
signal is sent to noflushd, it switches to the next set of timeout values in
the list, or jumps back to the first, if the end of the list was reached.
This is different from restarting the daemon with a new set of parameters
because the idle times for each disk are preserved.  You can make use of two
special timeout values with the
.B "-t "
option.  -1 or the keyword "default" uses the current default timeout.  0 or
the keyword "skip" advises noflushd to skip monitoring the particular disk
in this set of parameters.  Different timeout-lists may have different numbers
of entries, but this is probably not very useful at all.  Nevertheless it is
considered a feature, not a bug.  It is considered a bug however that this
description is far too confusing.
.TP 0.5i
.B "-i <irq list>"
Monitor interrupts for interactivity.  If one or more interrupts are raised
within the 
.I default
timeout, disks won't get spun down.  Interrupts are given
as a comma-separated list of either numbers or names as in
.IR /proc/interrupts.
You can use this option to prevent spindown when the machine is in interactive
use.  Keyboard, and mouse interrupts are good candidates to monitor in this
case.
.TP 0.5i
.B "-yes-scsi-idle-patch-is-in-kernel"
Enable support for SCSI disks.  The name of this option is gross for a reason:
Unless your running kernel has the scsi-idle patch applied, once spun down,
SCSI disks will not spin up again, no matter if noflushd is running or not.
(You might still be able to trigger a manual spinup, but by Murphy's Law the
program to do so resides on the spun-down disk.) Furthermore, the scsi-idle
patch is rumored to be buggy on multi-processor systems. Some people have
even blamed scsi-idle as the cause of occasional data loss on their disks.
In other words, by enabling this option, you assert that you know what you're
doing, and that you won't get at me if it turns out that in fact you didn't.
More information on SCSI support and the scsi-idle patch is provided in the
noflushd README, and in the contrib directory of the noflushd source
distribution.
.SH EXAMPLES
.TP 0.5i
.B 'noflushd -n 60 /dev/hda'
This is the recommended commandline for the most common laptop setup: one IDE
hard disk, optionally one IDE CD-ROM.  The CD-ROM is auto-detected as a
read-only drive.  Usually CD-ROM drives spin down themselves, so they should
not be given on the commandline for noflushd to monitor.
.TP 0.5i
.B 'noflushd -r /dev/sdb -n 60,5 /dev/hda -t 15,default /dev/hdc'
Here noflushd is told about a read-only device at /dev/sdb (e.g. a
magneto-optical drive mounted read-only).  noflushd needs this information to
decide what disks to sync and when to start/stop kupdated/pdflush.  Next, two
sets of timeout values are defined: First a 1 hour timeout for the first hard
disk, and a 15 minute timeout for the second hard disk.  Then a 5 minute
timeout for both the first and the second hard disk.  Sending SIGHUP to
noflushd will toggle between both parameter sets.
.TP 0.5i
.B 'noflushd -i keyboard,"PS/2 Mouse" -n 5 -t 60 /dev/hda'
Spin down the first IDE disk after one hour, but only if the machine hasn't
been used interactively in the past five minutes.
.SH LIMITATIONS
When running a devfs-enabled kernel, noflushd won't work unless devfs is
mounted.  Devfs need not be used, or mounted on /dev, but it needs to be
mounted somewhere.  That's because the kernel exposes different device names in
/proc/partitions, depending whether devfs is compiled in or not.
.PP
Kernels prior to version 2.3.99pre8 will only provide disk statistics for the
first and second IDE controller.  Thus noflushd can't support any more.
Improvements in later kernels curiously enough led to also the second IDE
controller not being supported anymore by default.  It's a kernel limitation,
not a noflushd bug.  If noflushd refuses to work with your UDMA-100 IDE disks,
this is most likely the cause.  Check noflushd's README in the doc section for
information on how to change this.
.PP
As if this hadn't been enough, 2.4.0-test9 changed the /proc/stat layout once
again. Current versions of noflushd only support the new format, meaning
that noflushd won't run on kernels 2.3.99pre8 to 2.4.0-test8.  Please consider
upgrading.
.SH BUGS
Heavy swap activity and spun down disks upon APM suspend sometimes leave
syslogd stuck in 'D' state after resuming. This is probably not a noflushd bug
but a kernel bug that can be triggered with the help of noflushd.  The best
known workaround is to put a line
.I "sync && sync && sync "
in apmd's suspend script.
.PP
Remounting a partition from read-only to read-write while noflushd's internal
syncer is active will go unnoticed, i.e. no sync will occur. (This should be
extremely rare - it requires that not only were e.g. /dev/hda{1,2...} all
read-only before, but also the full disk device /dev/hda. A filesystem on
an unpartitioned disk is probably the only real-world case where this bug
might trigger.)
.PP
Journaling filesystems like ext3, or reiserfs bypass the kernel's delayed
write mechanisms and write straight to disk.  Therefore noflushd is unable to
postpone writing of journaling data.  As a result, expect lousy spindown
behaviour when working off ext3/reiserfs/... partitions.
.PP
Handling of hotplugged drives is well below the optimum, especially when
removable drives like card readers are plugged in.  Until noflushd becomes
more clever about them, it is recommended to intelligently restart it
from the hotplug daemon (e.g. hotplug or usbmgr) as soon as a new drive
is added or removed.  Option -r might come in handy for the hotplugged
disk.
.PP
Some of the early 2.4 kernels apparently do not properly reap noflushd's
starting process after it has forked off the daemon: The entries in
/proc/<pid>/ are still around, but the pid doesn't show up anymore in the
directory contents of /proc/.  Which means chkrootkit reports noflushd as a
hidden process.  This bug is just so weird.  To daemonize itself, noflushd
uses a standard libc call, and there are no tricks in the source that should
be able to confuse the kernel.  Current kernels for all I know don't exhibit
this behaviour anymore.  Still why exactly noflushd triggered this bug remains
a mystery.
.PP
.SH AUTHOR
.B noflushd
was written by Daniel Kobras <kobras@linux.de>, based on
bdflush's sleep support.
.B bdflush
was written as a quick hack by Eric Youngdale <ericy@gnu.ai.mit.edu> with
various other hacks by Phil Bostley <bostley@cs.colorado.edu> and
Daniel Quinlan <quinlan@yggdrasil.com>.  Sleep support was added to bdflush
by Pavel Machek <pavel@atrey.karlin.mff.cuni.cz>.



noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



noflushd-2.8.netgear12/src/noflushd.c
/*
 * <noflushd.c>
 *
 * noflushd is a daemon to spin down disks after a certain amount of idle
 * time. It's useful on Linux systems of version 2.2.11 and up that have a
 * kernel thread named kupdate. For earlier versions use bdflush 1.6.
 *
 * noflushd version 1.7 was heavily based on bdflush 1.6 by 
 * Eric Youngdale (<ericy@gnu.ai.mit.edu>, 
 * Phil Bostley <bostley@cs.colorado.edu>,
 * Daniel Quinlan <quinlan@yggdrasil.com>, and
 * Pavel Machek <pavel@atrey.karlin.mff.cuni.cz>. 
 * Pavel Machek added sleep support to bdflush 1.6. noflushd still uses his
 * basic idea, but was completely rewritten in version 1.8 so the code 
 * doesn't bear much (if any) resemblance to bdflush anymore.
 *
 * noflushd is currently maintained by Daniel Kobras <kobras@linux.de>
 *
 * Copyright (C) 1999, 2000 Daniel Kobras
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * $Id: noflushd.c,v 1.22 2004/04/08 20:13:44 nold Exp $
 *
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include "disk_info.h"
#include "disk_stat.h"
#include "intr_stat.h"
#include "state.h"
#include "bug.h"
#include "pidfile.h"
#include "timeout.h"
#include "noflushd.h"

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <signal.h>
#include <syslog.h>
#include <time.h>
#include <string.h>
#include <errno.h>
#include <mntent.h>
#include <sys/mount.h>
#include <sys/stat.h>
#include <sys/wait.h>

extern int param_set(const char *path, int val); 

kupdate_t kupdate;
volatile int advance_timeout;
volatile int quitting;
static pidfile_t pid;

int verbose=0;
int released=0;
int staggered_spinup=0;
int block_dump=0;

extern int nfd_do_scsi;

static void init_sys_param(int enable)
{
	if (staggered_spinup) 
		param_set("/sys/module/scsi_mod/parameters/staggered_spinup", enable);
}

static void hup_handler(int i)
{
	DEBUG("Switching to next timeout set.");
	advance_timeout++;
}

static void term_handler(int i)
{
	INFO("Quitting on signal...");
	sync();
	advance_timeout++;
	quitting++;
	kill(getpid(), SIGUSR2);
}

#ifdef __GNUC__
static void usage(char **, int) __attribute__ ((noreturn));
#endif

static void usage(char **argv, int err)
{
	printf("Usage: %s [ -d | -v | -h | -n <timeout> ] [ -r <disk> ]\n"
	       "          [ -t <timeout> ] [-i <intrs>] <disk> ...\n", argv[0]);
	printf("  -h  Give this help\n");
	printf("  -V  Show version\n");
	printf("  -v  Be verbose(it can be used multiple times like -v -v to be more verbose)\n");
	printf("  -d  Show debugging information\n");
	printf("  -r  Mark disk as read-only\n");
	printf("  -n  Default timeout for spindown in minutes\n");
	printf("  -t  Spindown timeout in minutes for next disk\n");
	printf("  -i  Comma separated list of irqs to check interactivity\n");
	printf("  -b  enable block dump\n");
	exit(err);
}

int main(int argc, char *argv[])
{
	unsigned int i;
	unsigned int ch;
	int ret;
	int debug=0;
	
	struct stat statbuf;
	
	disk_stat_t diskstat;
	disk_info_t diskinfo, di;
	tolist_t next_timeout = NULL;

	unsigned int monitor = 0;
	int default_timeout=0;

	extern char *optarg;
	extern int optopt;

	char buffer[512];
	char model[17];
	char *irq = NULL;

	FILE *boot_info = fopen("/proc/sys/dev/boot/info", "r");
	if (boot_info) {
		fgets(buffer, sizeof(buffer)-1, boot_info);
		while (!feof(boot_info)) {
			if (sscanf(buffer, "model: %16[^\n]", model) == 1) {
				INFO("Found model: '%s'", model);
				if (strcmp(model, "ReadyNAS 104") == 0 ||
				    strcmp(model, "ReadyNAS 204") == 0 ||
				    strcmp(model, "ReadyNAS 214") == 0 ||
				    strcmp(model, "ReadyNAS 314") == 0)
					staggered_spinup = 1;
				break;
			}
			fgets(buffer, sizeof(buffer)-1, boot_info);
		}
		fclose(boot_info);
	} 
	else  {
		ERR("No Model Number");
	}

	/* We absolutely depend on /proc - check if it is online. */
	ret = stat("/proc/version", &statbuf);
	if (ret == -1) {
		perror("Checking /proc/version failed");
		ERR("Cannot access /proc - giving up");
		exit(1);
	}
	
	if (geteuid() != 0) {
		ERR("%.64s: not run as superuser", argv[0]);
		exit(1);
	}

	/* disk_info_init must be called before the cmdline
	 * is parsed, so for sane debugging, we have to parse several
	 * options by hand. (Yes, ain't pretty, but...)
	 */
	for (i=1; i < argc; i++) {
		if (argv[i][0] != '-')
			continue;
		if (argv[i][1] == 'd' && argv[i][2] == '\0') {
			debug++;
			verbose++;
			continue;
		}
		if (argv[i][1] == 'v' && argv[i][2] == '\0')
			verbose++;

		if (argv[i][1] == 'h' && argv[i][2] == '\0')
			usage(argv, 0);	/* usage() never returns */
	}
		
	devname_init();

	diskinfo = disk_info_init();
	if (!diskinfo)
		BUG("Disk info init failed.");

	if (!intr_stat_init())
		BUG("Interrupt stat init failed.");

	while ((ch = getopt(argc, argv, "-dvVhbxr:n:t:i:y:")) != EOF) {
		switch (ch) {
		case 'V':
			printf("%s - idle disk daemon. Version %s.\n",
			       PACKAGE, VERSION );
			break;
		case 'h':
		case 'd':
		case 'v':
			/* Help, debug and verbose were handled seperately. */
			break;
		case 'n':
			default_timeout=atoi(optarg)*60;
			timeout_parse(NULL, optarg, 1);
			next_timeout = NULL;
			break;
		case 'r':
			di=disk_info_get_byname(diskinfo,optarg);
			disk_info_mark_ro(di);
			break;
		case 't':
			next_timeout=timeout_parse(NULL, optarg, 0);
			break;
		case 'i':
			intr_stat_register_byids(optarg);
			break;
		case 'b':
			block_dump = 1;
			break;
		case 'y':
			if (!strcmp(optarg, "es-scsi-idle-patch-is-in-kernel"))
				nfd_do_scsi = 1;
			break;
		case 1:
			di=disk_info_get_byname(diskinfo,optarg);
			disk_info_setup(di, next_timeout);
			next_timeout = NULL;
			monitor++;
			break;
		case ':':
			ERR("Option -%c requires an argument", optopt);
			usage(argv, 1);	/* never returns */
		default:
			usage(argv, 1);	/* never returns */
		}
	}
	if (irq)
		intr_stat_register_byids(irq);

	for(i=optind; i < argc; i++) {
		di=disk_info_get_byname(diskinfo,argv[i]);
		disk_info_setup(di, next_timeout);
		next_timeout = NULL;	
		monitor++;
	}
	
	/* If no devices to monitor are given on the command line, try
	 * to find a sensible default. We monitor all read-write devices
	 * found; read-only devices are disregarding because in most cases
	 * they're CD-ROMs or DVDs that are not meant to spindown via
	 * noflushd.
	 */
	if (!monitor) {
		DEBUG("No devices given - autoprobing.");
		disk_info_foreach(diskinfo, di) {
			if (!disk_info_is_rw(di))
				continue;
			disk_info_setup(di, NULL);
			monitor++;
		}
	}

	if (!monitor) {
		INFO("No devices found - giving up.");
		exit(1);
	}
	
	kupdate = kupdate_init();
	if (!kupdate) {
		/* This can happen on pre-2.2.11 kernels. (Or if someone
		 * yet again decides to rename kupdate[d].) 
		 */
		INFO("No kupdate found. Giving up.");
		exit(1);
	}
		
	diskstat = disk_stat_init();
	if (!diskstat) {
		ERR("Disk stat init failed. Does your kernel provide "
		    "statistics for your drives?");
		exit(1);
	}
		
	intr_set_timeout(default_timeout);
	init_sys_param(1);
	
	if (debug||(!daemon(0, 0))) {
		if (debug) {
			signal(SIGINT, &term_handler);
		} else {
			released=1;
			openlog("noflushd", LOG_CONS | LOG_PID, LOG_DAEMON);
			signal(SIGINT, SIG_IGN);
		}
		signal(SIGHUP, &hup_handler);
		signal(SIGTERM, &term_handler);	/* must restart kupdate! */
		signal(SIGUSR2, SIG_IGN);

		if (!(pid = pidfile_get(NOFLUSHD_PID_FILE))) {
			ERR("Error creating pidfile");
			exit(1);
		}
		nfd_daemon(diskinfo, diskstat);

		if (!quitting)
			BUG("Huh!? Main daemon returned?");

		kupdate->start(kupdate);
		pidfile_drop(pid);
		unlink("/run/spindown");
		init_sys_param(0);
		adjust_vm_settings(0);
		exit(0);
	}
	ERR("Unable to fork off daemon: %s", strerror(errno));
	exit(1);
}
